[#1500] Add a Learn basics section to the DG (#1507)
reposense/RepoSense: docs/dg/learningBasics.md

1. In `Edit Configurations` of **Intellij**, set the current task to be the Gradle task `RepoSense [run]`, and click the debugging button on the right to start debugging (Make sure that you import the project as a Gradle project first according to [the tutorial](https://se-education.org/guides/tutorials/intellijImportGradleProject.html)).
1. Step into each break point when you think it is necessary to know more about the implementation, and step over otherwise to continue the execution.

<box type="info" seamless>

When tracing through the program execution, you can cross reference the architecture diagram and Javadoc of the class and method to check your understanding of the procedure.
</box>

{{ step(3) }} Get started with contributing

> Here are some small [tasks](https://github.com/reposense/RepoSense/issues?q=is%3Aissue+label%3Ad.FirstTimers+label%3Aa-Backend+is%3Aopen) for you to get started on contributing to RepoSense backend. Select an issue and try to work on it according to the guidance given in the issue comment.

Author: Acutally, will it be suitable to pin a specific first timer issue and advice the developer to just try it locally without submitting the pull request, or closing the pull request after submitting it in order for the developer to just gain some experience?

> Here are some small [tasks](https://github.com/reposense/RepoSense/issues?q=is%3Aissue+label%3Ad.FirstTimers+label%3Aa-Backend+is%3Aopen) for you to get started on contributing to RepoSense backend. Select an issue and try to work on it according to the guidance given in the issue comment.

Reviewer: These tasks need not be PRs. They can be as simple as modifying some UI element. They can be designed to expose the student to various important parts of the code base. The person should do them locally to self-test knowledge. If they encounter problems and need help, can push to the fork and ask for help by posting in our discussion board. We can give hints or even the model solution inside a panel so that the person can self-solve the problem without even bothering us.
BTW, we can also call them 'tutorials'.

> Here are some small [tasks](https://github.com/reposense/RepoSense/issues?q=is%3Aissue+label%3Ad.FirstTimers+label%3Aa-Backend+is%3Aopen) for you to get started on contributing to RepoSense backend. Select an issue and try to work on it according to the guidance given in the issue comment.

Author: Thanks for the suggestion. I will search for tasks along with hint and suggested solution and add them to the section.

> Here are some small [tasks](https://github.com/reposense/RepoSense/issues?q=is%3Aissue+label%3Ad.FirstTimers+label%3Aa-Backend+is%3Aopen) for you to get started on contributing to RepoSense backend. Select an issue and try to work on it according to the guidance given in the issue comment.

Reviewer: You can get help from @gerhean  and @dcshzj as well, because different members are more knowledgeable about different parts of the codebase.

> Here are some small [tasks](https://github.com/reposense/RepoSense/issues?q=is%3Aissue+label%3Ad.FirstTimers+label%3Aa-Backend+is%3Aopen) for you to get started on contributing to RepoSense backend. Select an issue and try to work on it according to the guidance given in the issue comment.

Author: Sure, no problem

<box type="info" seamless>

Some issues, such as [#937](https://github.com/reposense/RepoSense/issues/937), have linked pull requests from previous contributors. You can refer to these pull requests for implementation details.
</box>

<box type="warning" seamless>

If you want to try some more challenging tasks, you can check the issues in the [roadmap for contributors](https://github.com/reposense/RepoSense/projects/2#column-13522966).
</box>

{{ step(4) }} Get familiar with the workflow

> The next step is to enhance the code quality according to the coding standard, write necessary tests and documentation, and verify that your changes passes the test cases. After that, you can submit the pull request. You can refer to [the _Workflow_ page](workflow.html).

Reviewer: Step 5 can be doing actual PRs.

Here are some small [tasks](https://github.com/reposense/RepoSense/issues?q=is%3Aopen+is%3Aissue+label%3Aa-FrontEnd+label%3Ad.FirstTimers) for you to get started on contributing to RepoSense frontend. Select an issue and try to work on it according to the guidance given in the issue comment.

<box type="info" seamless>

Some issues, such as [#1205](https://github.com/reposense/RepoSense/issues/1205), have linked pull requests from previous contributors. You can refer to these pull requests for implementation details.
</box>

<box type="warning" seamless>

If you want to try some more challenging tasks, you can check the issues in the [roadmap for contributors](https://github.com/reposense/RepoSense/projects/2#column-13522966).
</box>

{{ step(4) }} Familiarize yourself with the workflow

The next step is to enhance the code quality according to the coding standard, write necessary tests and documentation, and verify that your changes passes the test cases. After that, you can submit the pull request. You can refer to [the _Workflow_ page](workflow.html).

<!-- ==================================================================================================== -->

## DevOps

> If you want to understand and contribute to the Continuous Integration aspect of RepoSense, you can refer to the [DevOps guide](https://github.com/reposense/RepoSense/wiki/DevOps-guide) for more information.

Reviewer: ```suggestion
If you want to understand and contribute to the DevOps aspect of RepoSense, you can refer to the [DevOps guide](https://github.com/reposense/RepoSense/wiki/DevOps-guide) for more information.
```

<variable name="title">Learning the Basics</variable>
<frontmatter>
  title: "{{ title | safe }}"
  pageNav: 3
</frontmatter>

{% from 'scripts/macros.njk' import embed, step with context %}

<h1 class="display-4"><md>{{ title }}</md></h1>

<div class="lead">

This is a learning guide for developers who are new to RepoSense.
</div>

<box type="warning" seamless>

Depending on what you have already known and what you would like to work on (back-end or frond-end), you may find certain sections irrelevant to you and you can skip them accordingly.
</box>

## back-end

<box type="info" seamless>

This section is for developers who want to contribute to the back-end of RepoSense. You may skip this section if you want to contribute as a pure front-end developer. Before you get started, you should have set up the project on your computer according to [the _Setting up_ page](settingUp.html).
</box>

The back-end implementation of RepoSense is located in `src/main`.

{{ step(1) }} Meet the language requirement

The RepoSense back-end is mostly written in `Java 8`

> 1. If you are not familiar with Java, you can refer to the [Java tutorial](https://docs.oracle.com/javase/tutorial/) to learn the basic syntax, standard library, and its object oriented nature. There are also plenty of other resources available from Youtube and other sites, and you should feel free to search for the learning resources that are most suitable to you.

Author: I think these are more advanced topics. I will put it in a separate step.

1. You may also find the [Java API](https://docs.oracle.com/javase/8/docs/api/) quite helpful when you are unsure about the usage of a certain class or method in a Java program.

{{ step(2) }} Learn the RepoSense back-end architecture

You may want to refer to the [back-end architecture](architecture.html). This can greatly help you to understand the implementation logic.

To gain a more concrete idea of how the backend works, you can use the IDE Debugger to trace through the steps of how the Json files are generated by taking the following steps.

<box type="warning" seamless>

The steps are for **Intellij**. If you are using a different IDE, you may need to check the documentation of how to use the debugger separately.
</box>

 ![tracing-using-debugger](../images/trace-using-debugger.jpg)

> *Figure 1. Tracing the execution using the debugger*

Reviewer 2: I think we don't need the caption here.

1. In `Edit Confiurations` of **Intellij**, set the current task to be the gradle task `RepoSense [run]`, and click the debugging button on the right to start debugging (Make sure that you import the project as a gradle project according to [the tutorial](https://se-education.org/guides/tutorials/intellijImportGradleProject.html)).
1. Step into each break point when you think it is necessary to know more about the implementation, and step over otherwise to continue the execution.

<box type="info" seamless>

When tracing through the program execution, you can cross reference the architecture diagram and `javadoc` of the class and method to check your understanding of the procedure.
</box>

{{ step(3) }} Get started with contributing

Here is a small [task](https://github.com/reposense/RepoSense/issues/1118) for you to get started on contributing to RepoSense back-end.

* The basic idea is to change or remove the validation for the author name tag located in `AnnotatorAnalyzer.java` by modifying the`REGEX_AUTHOR_NAME_FORMAT` argument.

> * You may need to check that the unit and system testing still passes after making the change by running the command `gradlew systemtest` and `gradlew test`.

Reviewer 2: ```suggestion
* You may need to check that the unit and system tests still pass after making the change, by running the commands `gradlew test` and `gradlew systemtest` respectively.
```

Here is more explanation about the usage of author tag.

> {{ embed("Appendix: **Using `@@author` tags**", "../ug/usingAuthorTags.md") }}

Author: Sure, no problem. I will add it to the issue

{{ step(4) }} Get familiar with the workflow

1. The next step is to verify that your changes passes the testing and then submit the pull request. You can refer to [the _Workflow_ page](workflow.html).

<!-- ==================================================================================================== -->

> ## front-end

Reviewer 2: Same here, we can put as "Frontend".

<box type="info" seamless>

> This section is for developers who want to contribute to the back-end of RepoSense. You may skip this section if you want to contribute as a pure front-end developer.

Reviewer 2: This section should be adapted since it is now talking about the frontend.

</box>

The front-end implementation of RepoSense is located in `frontend/src`.

{{ step(1) }} Learn the necessary tools

It is necessary for you to learn the basics of `Vue.js`, `pug`, and `scss` before working on the project. 

  {{ embed('Appendix: tools to learn for front-end development', 'tools.md', level=2) }}

> {{ step(3) }} Learn the RepoSense front-end architecture

Reviewer 2: The step number jumped, should be step 2.

* You may want to refer to the [front-end architecture](report.html#report-architecture) to understand the relationship

> * Another way for you to understand the front is to use **Vue.js devtools** to inspect the how the various Vue Components interact with each other. You can refer to the [front-end debugging guide](workflow.html#debugging-front-end) for more information.

Reviewer 2: "understand the front" and "inspect the how the various" seems a little weird. Need to reword the sentence.

> {{ step(4) }} Get started with contributing

Reviewer 2: Same for this, should be step 3.

> Here is a small [task](https://github.com/reposense/RepoSense/issues/1268) for you to get started on contributing to RepoSense front-end.

Reviewer 2: Same as for backend, we should use the issue tags to filter, so that we don't have to keep changing this link.

* The general goal is to eliminate the passing of variables that can be shared across different components, and move these variables to the **Vuex store**.
* You can try to move `minDate`, `maxDate`, and `filterSearch` to **Vuex store**. 

> * When in doubt, you can refer to the pull request [#1319](https://github.com/reposense/RepoSense/pull/1319) and [#1320](https://github.com/reposense/RepoSense/pull/1320) to view the implementation detail.

Reviewer 2: Let's move this into the issue itself with instructions on how and what to change.

{{ step(4) }} Familiarize yourself with the workflow

> 1. The next step is to verify that your changes passes the testing and then submit the pull request. You can refer to [the _Workflow_ page](workflow.html).

Reviewer 2: Same thing as for backend, I think we don't need a numbered list here.

Reviewer 2: ```suggestion
The next step is to verify that your changes passes the test cases before submitting the pull request. You can refer to [the _Workflow_ page](workflow.html).
```

<!-- ==================================================================================================== -->

## Writing documentation

If you want to make contribution to the RepoSense documentation, you can refer to the guide below.

**This project uses [MarkBind](https://markbind.org/)** for documentation. Follow [this tutorial](https://se-education.org/guides/tutorials/markbind.html) to learn how to use MarkBind for updating project documentation.

**To show some content only in the <tooltip content="i.e., https://reposense.org">production website</tooltip>**, add the `tags="production"` attribute to the HTML element enclosing the content. Similarly, `tags="dev"` will make the content appear only in the <tooltip content="i.e., https://reposense.org/RepoSense">dev website</tooltip>.

```html
<span tags="production">This will appear in the production website only.</span>
<span tags="dev">This will appear in the dev website only.</span>
This will appear in both sites.

> ```

Reviewer 2: I think this part should not be part of the basics page, since most (if not all) documentation should appear in both sites. Perhaps move it to other page that is more relevant for senior developers?

<!-- ==================================================================================================== -->

> ## Coding standard

Reviewer 2: Coding standards (plural)

* Make sure you know our coding standards.
  {{ embed('Appendix: Coding Standards', 'styleGuides.md', level=2) }}

> * **Follow [this](https://se-education.org/guides/tutorials/intellijCodeStyle.html) to configure Intellij to follow our coding style**.

Reviewer 2: No "this" or "click here" links.

* **This project uses Checkstyle** to check the compliance of Java code. You can use [this document](https://se-education.org/guides/tutorials/checkstyle.html) to find how to use it. In particular, run `gradlew checkstyleMain checkstyleTest checkstyleSystemtest` to check the style of all the relevant Java code.

> * **To check Pug files for style errors**, run `npm run lint` from the project root directory. You can use the `npm run lintfix` to automatically fix some of the javascript and css lint errors.

Reviewer 2: ```suggestion
* **To check Pug files for style errors**, run `npm run lint` from the project root directory. You can use the `npm run lintfix` to automatically fix some of the JavaScript and CSS lint errors.
```

<!-- ==================================================================================================== -->

## DevOps

> If you want to work on the Continuous integration of RepoSense, you can refer to the [DevOps guide](https://github.com/reposense/RepoSense/wiki/DevOps-guide) for more information.

Reviewer 2: ```suggestion
If you want to understand and contribute to the Continuous Integration aspect of RepoSense, you can refer to the [DevOps guide](https://github.com/reposense/RepoSense/wiki/DevOps-guide) for more information.
```

<variable name="title">Learning the Basics</variable>
<frontmatter>
  title: "{{ title | safe }}"
  pageNav: 3
</frontmatter>

{% from 'scripts/macros.njk' import embed, step with context %}

<h1 class="display-4"><md>{{ title }}</md></h1>

<div class="lead">

This is a learning guide for developers who are new to RepoSense.
</div>

<box type="warning" seamless>

> Depending on what you have already known and what you would like to work on (backend or frontend), you may find certain sections irrelevant to you and you can skip them accordingly.

Reviewer: ```suggestion
Depending on what you know already and what you would like to work on (i.e., backend or frontend), you may find certain sections irrelevant to you and you can skip them accordingly.
```

Author: Thanks for the suggestion. I will change it accordingly

</box>

## Backend

<box type="info" seamless>

This section is for developers who want to contribute to the backend of RepoSense. You may skip this section if you want to contribute as a pure frontend developer. Before you get started, you should have set up the project on your computer according to [the _Setting up_ page](settingUp.html).
</box>

The backend implementation of RepoSense is located in `src/main`.

> {{ step(1) }} Know Java

Reviewer: ```suggestion
{{ step(1) }} **Know Java**
```

Do the same for other similar locations.

</box>

* Check the [debugging guide](https://www.jetbrains.com/help/idea/debugging-your-first-java-application.html) if you are not familiar with debugging in **Intelij**.
* In `RepoSense.java`, the main class, set appropriate break points. Here are some relevant method calls in the `main` method at which you can set the breakpoints:
  * `ArgsParser.parse(args)`: RepoSense parses the CLI arguments from the command given by the user.
  * `getRepoConfigurations(cliArguments)`: RepoSense gets the configuration for each repository by parsing the CSV files.
  * `getReportConfigurations(cliArguments)`: RepoSense gets the report configuration (report title) by parsing the JSON files.
  * `ReportGenerator.generateReposReport(...)` This is where the bulk of the work is conducted, including cloning repositories, analyzing repositories, and generating the JSON files to be used by the report.

<box type="info" seamless>

When tracing through the program execution, you can cross reference the architecture diagram and Javadoc of the class and method to check your understanding of the procedure.
</box>

{{ step(3) }} Gain some hands-on experience

> Here are some small tasks for you to get started on RepoSense backend contribution.

Reviewer: ```suggestion
Here are some small tasks for you to gain some basic knowledge of the code related to the RepoSense backend. You can do each in a separate branch in your local copy of the code.
```

> <panel header="**Command Line Argument Parser**" type="primary">

Reviewer: ```suggestion
<panel header="**Task 1: Add a flag to pretty-print the JSON file**" type="primary">
```

<panel header="Hint 1">

Try to understand the parsing process of the Command Line Arguments, which starts from `ArgsParser.parse(args)` in `RepoSense.java` and uses `ArgumentParser` to convert an argument string into a Java Object.

Therefore, the first step you can take is to add the following to `ArgsParser`.

```
public static final String[] JSON_PRINT_MODE_FLAGS = new String[]{"--use-json-pretty-printing", "-j"};
```

In `getArgumentParser` method, add the following content.

```
parser.addArgument(JSON_PRINT_MODE_FLAGS)
    .dest(JSON_PRINT_MODE_FLAGS[0])
    .action(Arguments.storeTrue())
    .help("A flag to use json pretty printing when generating the json files.");
```
</panel>

<panel header="Hint 2">

After the step in hint 1, the argument is captured by `ArgumentParser`. Now change the  `CliArguments`, `ConfigCliArguments`, `parse` method in `ArgsParser` to make the return result of `parse` include the new argument. 

Add the following content to `CliArguments`.

```
protected boolean isPrettyPrintingUsed;

public boolean isPrettyPrintingUsed() {
        return isPrettyPrintingUsed;
}
```

> In the [constructor](https://github.com/reposense/RepoSense/blob/9125d4b6bea1e5bb6329cb8f7c476ea18fed3cea/src/main/java/reposense/model/ConfigCliArguments.java#L26-L52) of `ConfigCliArguments`, add `isPrettyPrintingUsed` as a new argument of the method, add the following content to the method body.

Reviewer: Perhaps not fix the link to specific version or specific lines? If you do, it can get out of sync with the actual code sooner or later.

Author: Sure, no problem

```
this.isPrettyPrintingUsed = isPrettyPrintingUsed;
```

In the `parse` method, replace the [section](https://github.com/reposense/RepoSense/blob/9125d4b6bea1e5bb6329cb8f7c476ea18fed3cea/src/main/java/reposense/parser/ArgsParser.java#L303-L306) with the following content to get `isJsonPrettyPrintingUsed` from `ArgmentParser` and pass it to `ConfigCliArguments`.

```
boolean isJsonPrettyPrintingUsed = results.get(JSON_PRINT_MODE_FLAGS[0]);

return new ConfigCliArguments(configFolderPath, outputFolderPath, assetsFolderPath, sinceDate, untilDate,
        isSinceDateProvided, isUntilDateProvided, numCloningThreads, numAnalysisThreads, formats,
        shouldIncludeLastModifiedDate, shouldPerformShallowCloning, isAutomaticallyLaunching,
        isStandaloneConfigIgnored, zoneId, isJsonPrettyPrintingUsed);

> ``` 

Reviewer: Try to minimize giving the exact code, to reduce the risk of this document going out of sync with the actual code.

Author: Thanks for the suggestion. I will try to eliminate the code replacement and restrict the presence of such code section only when it is a pure addition

```
GsonBuilder gsonBuilder = new GsonBuilder()
        .setDateFormat(GITHUB_API_DATE_FORMAT)
        .registerTypeAdapter(FileType.class, new FileType.FileTypeSerializer());
Gson gson;
if (isPrettyPrintingUsed) {
    gson = gsonBuilder.setPrettyPrinting().create();
} else {
    gson = gsonBuilder.create();
}
```

To notify `FileUtil` of the switch between different printing mode, add the following content to `FileUtil`.

```
public static void setPrettyPrintingMode(boolean isPrettyPrintingAdopted) {
    isPrettyPrintingUsed = isPrettyPrintingAdopted;
}
```

It is now possible to notify `FileUtil` of the printing mode switch by adding the following content to the `main` method of `RepoSense.java`.

```
FileUtil.setPrettyPrintingMode(cliArguments.isPrettyPrintingUsed());
```

Now the parsing of argument and changing of printing mode should have been completed.
</panel>

<panel header="Suggested solution">
  There is more than 1 way to achieve this. One solution is shown as the following:
  

>   According to hint 1, hint 2, and hint 3, here are the complete [required changes](https://gist.github.com/HCY123902/e85ad0f7b258937945ce1aa97fe4af9e/revisions) (See the latest revision).

Reviewer: Similarly, try to avoid giving exact code. Unlike AB3, RepoSense code is expected to evolve all the time --> the risk of code solutions can go out of sync pretty fast.

Author: Sure, no problem

</panel>

<panel header="Hint 3">

Check what `selectedCommits` does and how the link and commit title of each commit is retrieved.
</panel>

<panel header="Suggested solution">
  There is more than 1 way to achieve this. One solution is shown as the following:
  
  In the [indivdual commit section](https://github.com/reposense/RepoSense/blob/9125d4b6bea1e5bb6329cb8f7c476ea18fed3cea/frontend/src/tabs/zoom.pug#L88-L98) of `zoom.pug`, replace line 96 to 98 with the following [content](https://gist.github.com/HCY123902/e505429c26330df3ee17b1cb69972302/revisions) (See the latest revision).
</panel>

</panel>

<box type="info" seamless>

This is only for your practice. There is no need for you to commit this change and submit it in a pull request.
</box>

{{ step(4) }} Get familiar with the workflow

The next step is to learn how to enhance the code quality according to the coding standard, write necessary tests and documentation, verify that the changes passes the test cases, and submit the pull request. You can refer to [the _Workflow_ page](workflow.html).

{{ step(5) }} Get started with contributing

> RepoSense currently has a lot of open issues, and some may be challenging. Now that you are ready, you can check the [issues for first timers](https://github.com/reposense/RepoSense/issues?q=is%3Aopen+is%3Aissue+label%3Ad.FirstTimers+label%3Aa-FrontEnd) or issues in the [roadmap for contributors](https://github.com/reposense/RepoSense/projects/2#column-13522966) to work on actual pull requests.

Reviewer: These two steps should not be here as they are not part of 'learning the basics'. Those two steps are in a different page. You can add more details to that page if you want.

Author: Thanks for the suggestion. I will move it to the workflow page accordingly.
